doc/design/model_zoo_dev.md

```diff
@@ -0,0 +1,205 @@
+
+Companies including Ant Financial and DiDi deployed SQLFlow and
+integrated into the corp data infrastructure.  Developers in these
+companies don't need to install any software, but can use the Web GUI
+of their data infrastructure.
+
+## Model Development in Python
+
+### Define a Model
+
+For clarify, let us consider a deep learning model definition is a
+Keras class in Python.  For example, the following class in
+`my_keras_model.py`:
+
+```python
+import sqlflow
+
+class MyKerasModel(tf.keras.Model):
+   def __init__(self, ...):
+      ...
+```
+
+For detailed instructions, please refer to this
+[document](https://github.com/sql-machine-learning/models/blob/develop/doc/customized%2Bmodel.md).
+
+### Test and Verify
+
+To verify that this model deifniton works, we can call Keras
+high-level API like `fit` to train a model and then evaluate it.
+
+After that, we want to verify it works with SQLFlow.  This depends on
+the SQLFlow high-level Python API, which is a Python abstraction of
+the SQL syntax extension introduced by SQLFlow.
+
+### SQLFlow High-level API
+
+For example, in SQLFlow we can write
+
+```sql
+SELECT ... TO TRAIN model_def WITH ... COLUMNS ... INTO trained_model;
+```
+
+Accordingly, in Python, we can write
+
+```python
+import sqlflow
+sqlflow.train("SELECT ...",
+              model_def,
+              with={"param1":value, "param2":value2},
+			  columns={"column1":lambda(x):..., "column2":labmda(x):...},
+			  into="trained_model")
+```
+
+### Unit Tests
+
+With the high-level API, we can test the model definition by writing a
+unit test file like `test_my_keras_model.py`.
+
+```python
+import sqlflow
+
+def test_MyKerasModel():
+  ## 初始化。dbhost：本地数据库地址；AI引擎默认是本地 SQLFlow server 中的TensorFlow, XGBoost等
+  sqlflow.init(dbhost="mysql://127.0.0.1:3306")
+  sqlflow.train("SELECT ...", MyKerasModel, ... into="my_trained_model")
+  auc, f1 = sqlflow.evaluate("SELECT ...", using="my_trained_model", metric=["AUC", "F1"])
+  unittest.assert(auc > 0.9)
+  unittest.assert(f1 > 0.5)
+```
+
+The above exmaple connects to a MySQL database system.  We can connect
+to other supported database systems as well.
+
+```python
+import sqlflow
+
+def test_MyKerasModel():
+  ## 初始化SQLFlow。 project： maxcompute project；AI引擎默认是PAI
+  sqlflow.init(dbhost="maxcompute://[AK]:[SK]@service-corp.odps.aliyun-inc.com/api?curr_project=alifin_jtest_dev&scheme=http")
+  sqlflow.train("SELECT ...", MyKerasModel, ... into="my_trained_model")
+  auc, f1 = sqlflow.evaluate("SELECT ...", using="my_trained_model", metric=["AUC", "F1"])
+  unittest.assert(auc > 0.9)
+  unittest.assert(f1 > 0.5)
+```
+
+We can run these unit tests by typing the following command.
+
+```bash
+pytest
+```
+
+### Test Using SQLFlow Command-line Tool
+
+Like MySQL has a command-line client `mysql`, we provide `sqlflow`.
+Using this command-line tool, we can run the following command to use
+the model definition.
+
+```bash
+$ sqlflow
+sqlflow > SELECT * from iris
+TO TRAIN MyKerasModel
```

Comment:
After the above steps, the model developer has a .py file on her local computer including the model definition `class MyKerasModel`. How does `sqlflow` recognize the existence of the model definition and be able to use it?

Reply:
We can use Python reflection to mimic a duck typing mechanism to check the classes and functions in the python file, e.g. classes that have the expected interface, classes that derived from `Keras.Model`, or functions with the expected signature will be recognized as a model.